Add section on string coercion behavior (#20234)
mdn/content: files/en-us/web/javascript/reference/global_objects/string/index.md

console.log(eval(s2.valueOf()))  // returns the number 4
```

> ### String coercion

Author: The reason why I mentioned `ToString()` is because it's a baseline and it's mentioned multiple times downstream. Avoiding it makes it much harder to compare template literals, built-in coercion, and `String()`. Is it fine if I just add a caveat?

- Strings are returned as-is.
- [`undefined`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined) turns into `"undefined"`.
- [`null`](/en-US/docs/Web/JavaScript/Reference/Operators/null) turns into `"null"`.
- `true` turns into `"true"`; `false` turns into `"false"`.
- Numbers are converted with the same algorithm as [`toString(10)`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toString).
- [BigInts](/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) are converted with the same algorithm as [`toString(10)`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toString).
- [Symbols](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) throw a {{jsxref("TypeError")}}.
- Objects are first converted to a primitive by calling its [`[@@toPrimitive]()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive) (with `"string"` as hint), `toString()`, and `valueOf()` methods, in that order. The resulting primitive is then converted to a string.

> There are several ways to achieve nearly the same effect in JavaScript.
> 
> - [Template literal](/en-US/docs/Web/JavaScript/Reference/Template_literals): `` `${x}` ``
> - The [`String()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/String) function: `String(x)`
> - Using the [`+` operator](/en-US/docs/Web/JavaScript/Reference/Operators/Addition): `"" + x`
> 
> However, they do slightly different things.
> 
> - Template literals do exactly `ToString(x)` for the embedded expression.
> - The `String()` function does `ToString(x)`, except that [Symbols](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) don't throw a {{jsxref("TypeError")}}, but return `"Symbol(description)"`, where `description` is the [description](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/description) of the Symbol.
> - The `+` operator coerces its operand to a _primitive_ instead of a _string_, and, for some objects, has entirely different behaviors from `ToString`. See its [reference page](/en-US/docs/Web/JavaScript/Reference/Operators/Addition) for more details.

Reviewer: ```suggestion
To achieve the same effect in JavaScript, you can use:

- [Template literal](/en-US/docs/Web/JavaScript/Reference/Template_literals): `` `${x}` `` that do string coercion on the embedded expression
- The [`String()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/String) function: `String(x)`, except that [Symbols](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) don't throw a {{jsxref("TypeError")}}, but return `"Symbol(description)"`, where `description` is the [description](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/description) of the Symbol.
- Using the [`+` operator](/en-US/docs/Web/JavaScript/Reference/Operators/Addition): `"" + x` except that it coerces its operand to a _primitive_ instead of a _string_, and, for some objects, has an entirely different behavior. Refer to its [reference page](/en-US/docs/Web/JavaScript/Reference/Operators/Addition) for more details.
```
